### ISSUE TITLE
[BUG] save/load checkpoint in zero3 fails to preserve frozen weights

### ISSUE THREAD

**Issue by stas00** [microsoft/DeepSpeed#3090]
**Describe the bug**

I was scratching my head on why I wasn't getting on resume the same weights as what was going in during model save.

The problem stems from having some weights that are frozen. The checkpoint saves only optimizer states which don't have frozen weights in it! so those weights are lost.

So when we resume - all the frozen weights aren't there and remain random. This would be extremely subtle to catch if only a few weights are frozen.

I validated that if I unfreeze everything, the resume works just fine - what goes in - it comes out correctly
so basically we need to figure out what to do with frozen weights, so that they survive resume. 

Using `stage3_gather_16bit_weights_on_model_save` isn't an option for LLM. And even if there was enough cpu memory it'd be extremely inefficient if needing to save checkpoints frequently during training.

@tjruwase

**Comment by stas00** [comment]
While at it we need to raise priority for solving https://github.com/microsoft/DeepSpeed/issues/1971 since it again is a show stopper for using ZeRO3 for LLM training.

**Comment by tjruwase** [comment]
**TODOs**

1. Fix tests/unit/checkpoint/common.py::compare_model_states
 -> params_to_fetch = [p for p in loaded_model.params()+trained_model.params() if p.ds_status == NOT_AVAIL] 
    with Gather(params_to_fetch): 
      for x,y in zip(loaded_model.params(), trained_model.params()):
       
2. Flush z3 cache in following cases
     1. After training
     2. After ds.init()

3. Frozen weight tests
 -> Create 2 models with 1 frozen weight
  -> test_load_module_only
  -> test full deepspeed state
 
4. Checkpointing frozen weights
 -> Similar to param_shapes in module checkpoint
 -> Save param with requires_grad==False into module.state_dict
 -> _get_zero_param_shapes -> _get_zero_param_ds_tensor: Refactoring to generalize 
 -> Add frozen_param_groups to zero 1/2/3 optimizer state, mirror self.fp16_groups

5. Update zero_to_fp32.py to recover frozen weights.

**Comment by suri-kunal** [comment]
Hi @stas00 I am thinking about using ZeRO3 for training LoRA method in peft library from Huggingface. Since LoRA freezes the original model and adds an additional matrix on top of it, could this issue create a problem while training LoRA? Thanks

**Comment by stas00** [comment]
I shouldn't impact training itself.

This issue is about saving/loading checkpoint if you are not using the `"stage3_gather_16bit_weights_on_model_save": true,` setting. It's possible to work around those, but this is not trivial.

**Comment by bigeagle** [comment]
My current solution to save zero3 weights:
```python
    from deepspeed.utils import safe_get_full_fp32_param
    output_state_dict = {
        k: safe_get_full_fp32_param(v).cpu() for k, v in model.named_parameters()
    }

    if get_local_rank() != 0:
        return

    torch.save(output_state_dict, os.path.join(save_directory, WEIGHTS_NAME))
```

**Comment by tjruwase** [comment]
@bigeagle, @stas00, apologies for the delay on this. There should be some progress here by next week.

**Comment by suri-kunal** [comment]
I am using deepspeed_engine.save_checkpoint and deepspeed_engine.load_checkpoint to save and load models. Is it wrong? Also, I am not using ZeRO3 but ZeRO2 because I am getting some error of 'Index out of bounds` in ZeRO3

**Comment by suri-kunal** [comment]
> I shouldn't impact training itself.
> 
> This issue is about saving/loading checkpoint if you are not using the `"stage3_gather_16bit_weights_on_model_save": true,` setting. It's possible to work around those, but this is not trivial.

Thanks for your response. I am using deepspeed_engine.save_checkpoint and deepspeed_engine.load_checkpoint to save and load models. Is it wrong? Also, I am not using ZeRO3 but ZeRO2 because I am getting some error of 'Index out of bounds` in ZeRO3

**Comment by stas00** [comment]
> Thanks for your response. I am using deepspeed_engine.save_checkpoint and deepspeed_engine.load_checkpoint to save and load models. Is it wrong? 

That is the correct way in general. Except in the case of frozen weights, it won't save/restore frozen weights. So until this bug is fixed on resume you need to restore the frozen weights from the original ones (as they have never changed). Once it's fixed then it'll work w/o any workarounds.

> Also, I am not using ZeRO3 but ZeRO2 because I am getting some error of `Index out of bounds` in ZeRO3

Please file a new issue about it, including a full traceback. Most of the time this type of error comes a mismatch of embedding size and token ids, but if it works with Z2 then it's probably something else.

**Comment by richardwth** [comment]
> Hi @stas00 I am thinking about using ZeRO3 for training LoRA method in peft library from Huggingface. Since LoRA freezes the original model and adds an additional matrix on top of it, could this issue create a problem while training LoRA? Thanks

So is there a way to not save the frozen weights? I am using LoRA and the huge model got saved.

**Comment by tjruwase** [comment]
@richardwth, can you please add more details of the problem? It might be better to create a new ticket since it seems different from this one. Thanks!
